server/src/main/java/org/elasticsearch/index/reindex/ReindexTask.java

```diff
@@ -0,0 +1,140 @@
+
+    public static class ReindexPersistentTasksExecutor extends PersistentTasksExecutor<ReindexJob> {
+
+        private final ClusterService clusterService;
+        private final Client client;
+
+        public ReindexPersistentTasksExecutor(ClusterService clusterService, final Client client) {
+            super(NAME, ThreadPool.Names.GENERIC);
+            this.clusterService = clusterService;
+            this.client = client;
+        }
+
+        @Override
+        protected void nodeOperation(AllocatedPersistentTask task, ReindexJob reindexJob, PersistentTaskState state) {
+            ReindexTask reindexTask = (ReindexTask) task;
+            reindexTask.doReindex(reindexJob);
+
+        }
+
+        @Override
+        protected AllocatedPersistentTask createTask(long id, String type, String action, TaskId parentTaskId,
+                                                     PersistentTasksCustomMetaData.PersistentTask<ReindexJob> taskInProgress,
+                                                     Map<String, String> headers) {
+            return new ReindexTask(id, type, action, parentTaskId, headers, clusterService, client);
+        }
+    }
+
+    private ReindexTask(long id, String type, String action, TaskId parentTask, Map<String, String> headers,
+                        ClusterService clusterService, Client client) {
+        super(id, type, action, "reindex_" + id, parentTask, headers);
+        this.client = new ParentTaskAssigningClient(client, new TaskId(clusterService.localNode().getId(), id));
+    }
+
+    private void doReindex(ReindexJob reindexJob) {
+        ThreadContext threadContext = client.threadPool().getThreadContext();
+        // TODO: What is happening here? Putting headers back in place for possible different thread action listener?
+        final Supplier<ThreadContext.StoredContext> supplier = threadContext.newRestorableContext(false);
+        try (ThreadContext.StoredContext ignore = stashWithHeaders(threadContext, reindexJob.getHeaders())) {
+            // TODO: Always store currently for compatibility
+            ReindexRequest reindexRequest = reindexJob.getReindexRequest();
+            client.execute(ReindexAction.INSTANCE, reindexRequest, new ContextPreservingActionListener<>(supplier,
+                new ActionListener<>() {
+                    @Override
+                    public void onResponse(BulkByScrollResponse response) {
+                        updatePersistentTaskState(new ReindexJobState(response, null), new ActionListener<>() {
+                            @Override
+                            public void onResponse(PersistentTasksCustomMetaData.PersistentTask<?> persistentTask) {
+                                markAsCompleted();
+                            }
+
+                            @Override
+                            public void onFailure(Exception e) {
+                                logger.error("Failed to update task state to success.", e);
```

Comment:
Let's not have error logging please. Instead I think we will need to make updates to persistent tasks resilient against master failovers. Currently any updates can fail if there's a master failover or if there is a short disconnect between this node and the master.

Let's add this as a TODO to the meta issue